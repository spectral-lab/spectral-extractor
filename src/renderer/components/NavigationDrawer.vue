<template>
  <div>
    <v-navigation-drawer
      v-model="showDrawer"
      dark
      temporary
      absolute
    >
      <div
        :style="{height: `${parseInt(titleBarHeight) - 4}px`}"
        class="spacer"
      />
      <v-list dense>
        <v-list-group
          v-for="menuItem in menuTemplate"
          :key="menuItem.label"
          no-action
        >
          <template #activator>
            <v-list-tile>
              <v-list-tile-title>{{ menuItem.label }}</v-list-tile-title>
            </v-list-tile>
          </template>
          <v-card color="grey darken-2">
            <v-list-tile
              v-for="subItem in menuItem.subMenu"
              :key="subItem"
              @click="handleClick(subItem)"
            >
              <v-list-tile-title>
                {{ subItem }}
              </v-list-tile-title>
            </v-list-tile>
          </v-card>
        </v-list-group>
      </v-list>
      <div class="spacer" />
    </v-navigation-drawer>
  </div>
</template>

<script>
import { titleBarHeight } from '../constants/layout';

export default {
  props: {
    menuTemplate: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      showDrawer: false,
      titleBarHeight
    };
  },
  mounted () {
    this.listenMouseMove();
  },
  methods: {
    handleClick (type) {
      this.$emit('click', type);
    },
    hide () {
      if (this.showDrawer) this.showDrawer = false;
    },
    listenMouseMove () {
      //  If the mouse position keeps near the left edge 300 ms or longer, "showDrawer" is set true.
      let isNearLeftEdge = false;
      let waiting = false;
      document.addEventListener('mousemove', (ev) => {
        isNearLeftEdge = ev.pageX < 60;
        if (isNearLeftEdge) {
          if (!waiting) {
            setTimeout(() => {
              if (isNearLeftEdge) this.showDrawer = true;
              waiting = false;
            }, 300);
            waiting = true;
          }
        }
      });
      document.addEventListener('mouseleave', () => {
        isNearLeftEdge = false;
      });
    }
  }
};
</script>

<style scoped>
  .spacer {
    height: 50px;
  }
</style>
