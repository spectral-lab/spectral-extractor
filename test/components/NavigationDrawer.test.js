import { shallowMount } from '@vue/test-utils';
import NavigationDrawer from '../../src/renderer/components/NavigationDrawer';
import menuTemplate from '../../src/renderer/constants/menu';
import Vue from 'vue';
import vuetify from 'vuetify';

Vue.use(vuetify);

describe('Component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(NavigationDrawer, {
      propsData: {
        menuTemplate
      }
    });
  });
  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });
  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
