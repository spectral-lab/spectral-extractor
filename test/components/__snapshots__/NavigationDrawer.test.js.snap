// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly 1`] = `
<div>
  <v-navigation-drawer-stub
    absolute="true"
    dark="true"
    height="100%"
    minivariantwidth="80"
    mobilebreakpoint="1264"
    temporary="true"
    width="300"
  >
    <div
      class="spacer"
      style="height: 20px;"
    />
     
    <v-list-stub
      dense="true"
    >
      <v-list-group-stub
        activeclass="primary--text"
        appendicon="$vuetify.icons.expand"
        noaction="true"
      >
         
        <v-card-stub
          color="grey darken-2"
          tag="div"
        >
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              New Project
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              Open Project
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              Save Project
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              Export MIDI
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              Export Audio
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
        </v-card-stub>
      </v-list-group-stub>
      <v-list-group-stub
        activeclass="primary--text"
        appendicon="$vuetify.icons.expand"
        noaction="true"
      >
         
        <v-card-stub
          color="grey darken-2"
          tag="div"
        >
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              Undo
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
          <v-list-tile-stub
            activeclass="primary--text"
          >
            <v-list-tile-title-stub>
              
              Redo
            
            </v-list-tile-title-stub>
          </v-list-tile-stub>
        </v-card-stub>
      </v-list-group-stub>
    </v-list-stub>
     
    <div
      class="spacer"
    />
  </v-navigation-drawer-stub>
</div>
`;
